### PR TITLE
torch ranger 

### PR DESCRIPTION
**Pull request by fantes**
this PR adds ranger + extensions as native torch optimizer (ie coded in c++)
ranger is rectified ADAM  + lookahead
extensions added are : adabelief and centralized gradient
see docs/api and unit test for examples

### CODE REVIEW

File: src/backends/torch/torchmodule.cc [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+
+namespace dd
+{
+  // ======= TORCH MODULE
+
+  TorchModule::TorchModule() : _device{ "cpu" }
+  {
+  }
+
+  torch::Tensor TorchModule::to_tensor_safe(const c10::IValue &value)
```
Comment (BynaryCobweb) [inline src/backends/torch/torchmodule.cc:14]: This function is not meant to be part of TorchModule, it's just an utility function to safely extract a tensor from an IValue. Put it in another file (`torchutil.h`)?

File: src/backends/torch/torchlib.cc [hunk 2]
```diff
@@ -75,21 +74,6 @@ namespace dd
 #endif
   }
 
-  void add_parameters(std::shared_ptr<torch::jit::script::Module> module,
```
Comment (BynaryCobweb) [inline src/backends/torch/torchlib.cc:78]: The following functions are utility functions and should be moved in a separate header, e.g. `torchutil.h` so that they can be accessed by TorchModule, TorchSolver, etc

File: src/backends/torch/torchmodule.cc (hunk 1 above)
Comment (fantes) [reply src/backends/torch/torchmodule.cc:14]: done

File: src/backends/torch/torchlib.cc (hunk 2 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:78]: done

File: docs/api.md [hunk 3]
```diff
@@ -752,7 +752,16 @@ Parameter     | Type   | Optional | Default | Description
 ---------     | ----   | -------- | ------- | -----------
 iterations    | int    | yes      | N/A     | Max number of solver's iterations
 snapshot      | int    | yes      | N/A     | Iterations between model snapshots
-solver_type   | string | yes      | SGD     | from "SGD", "ADAGRAD",  "RMSPROP", "ADAM"
+solver_type   | string | yes      | SGD     | from "SGD", "ADAGRAD",  "RMSPROP", "ADAM", "RANGER"
```
Comment (beniz) [inline docs/api.md:755]: we may want RANGER_PLUS to make the API identical to that of https://github.com/jolibrain/deepdetect/pull/901

File: src/backends/torch/optim/ranger.cc [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+
+  void RangerParamState::serialize(torch::serialize::InputArchive &archive)
+  {
+    _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(int64_t, step);
+    _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(torch::Tensor, exp_avg);
+    _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(torch::Tensor, exp_avg_sq);
+    _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(torch::Tensor, slow_buffer);
+  }
+
+  torch::Tensor Ranger::step(LossClosure closure)
```
Comment (beniz) [inline src/backends/torch/optim/ranger.cc:87]: If it's inspired from libtorch Adam implementation for instance, I'd put a link to the original code, just in case if things change one day.

File: src/backends/torch/optim/ranger.cc [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+              }
+            else
+              {
+                auto N_sma_max = 2.0 / (1.0 - beta2) - 1.0; // rho_inf
+                auto beta2_t = std::pow(beta2, state.step());
+                auto N_sma
+                    = N_sma_max
+                      - 2.0 * (float)state.step() * beta2_t / bias_correction2;
+                double step_size;
+                if (N_sma >= 5)
```
Comment (beniz) [inline src/backends/torch/optim/ranger.cc:177]: FYI this value appears to differ from the one that you use with Caffe https://github.com/jolibrain/caffe/blob/6c9b6bee92a2825d50be07c8a9e523ed56bfef20/src/caffe/solvers/adam_solver.cpp#L110

File: src/backends/torch/optim/ranger.h [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+{
+  struct TORCH_API RangerOptions
+      : public torch::optim::OptimizerCloneableOptions<RangerOptions>
+  {
+    RangerOptions(double lr = 1e-3);
+    TORCH_ARG(double, lr) = 1e-3;
+    typedef std::tuple<double, double> betas_t;
+    TORCH_ARG(betas_t, betas) = std::make_tuple(0.9, 0.999);
+    TORCH_ARG(double, eps) = 1e-8;
+    TORCH_ARG(double, weight_decay) = 0.0;
```
Comment (beniz) [inline src/backends/torch/optim/ranger.h:31]: Just double checking: are we sure `weight_decay` is set to 0.0 by default ?

File: src/backends/torch/torchUtils.h [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+#include <torch/torch.h>
+#pragma GCC diagnostic pop
+#include <torch/script.h>
+
+namespace dd
+{
+
+  namespace torch_utils
+  {
+    inline void empty_cuda_cache()
```
Comment (beniz) [inline src/backends/torch/torchUtils.h:22]: This appears unused ? Also, we may need a warning comment on functions that affect the global CUDA state.

File: src/backends/torch/torchUtils.h [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    {
+#if !defined(CPU_ONLY)
+      c10::cuda::CUDACachingAllocator::emptyCache();
```
Comment (beniz) [inline src/backends/torch/torchUtils.h:25]: I know this is was not added by this PR, but beware of this, as my understanding is that it affects the global CUDA state. Do we know whether this is compatible with using multiple libtorch services ? cc @BynaryCobweb

File: src/backends/torch/torchmodule.h [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    void load(TorchModel &model);
+
+    void eval();
+    void train();
+
+    void free();
+
+    static torch::Tensor to_tensor_safe(const c10::IValue &value);
+
+    template <class TInputConnectorStrategy>
```
Comment (beniz) [inline src/backends/torch/torchmodule.h:53]: We need comments on these functions :)

File: src/backends/torch/torchlib.cc [hunk 10]
```diff
@@ -998,30 +467,13 @@ namespace dd
...
-            new optim::SGD(_module.parameters(), optim::SGDOptions(base_lr)));
-      }
+    tsolver.create(_module);
 
     int it = 0;
     // reload solver and set it value accordingly
-    solver_load(optimizer);
-    optimizer->zero_grad();
+    it = tsolver.load(this->_mlmodel._sstate);
+    // solver_load(optimizer);
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:475]: remove comment ?

File: src/backends/torch/torchUtils.h (hunk 8 above)
Comment (BynaryCobweb) [reply src/backends/torch/torchUtils.h:25]: This method is thread safe ([code](https://github.com/pytorch/pytorch/blob/8357e2edc329b3a62979ceae3faa1e51fd70150e/c10/cuda/CUDACachingAllocator.cpp#L377)) and only free unused memory, so I think using it is not a problem.

**Review by BynaryCobweb — APPROVED** [review]

File: src/backends/torch/optim/ranger.cc (hunk 4 above)
Comment (fantes) [reply src/backends/torch/optim/ranger.cc:87]: ok

File: src/backends/torch/optim/ranger.cc (hunk 5 above)
Comment (fantes) [reply src/backends/torch/optim/ranger.cc:177]: not really , this one states `>=5`, while the other states `>4` and N_sma is supposed to be an int

File: src/backends/torch/optim/ranger.h (hunk 6 above)
Comment (fantes) [reply src/backends/torch/optim/ranger.h:31]: default is no weight_decay in almost all optimizer implementation I've seen

File: src/backends/torch/torchlib.cc (hunk 10 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:475]: ok done

File: src/backends/torch/torchlib.cc [hunk 11]
```diff
@@ -998,30 +530,13 @@ namespace dd
...
```
[the remark this answers is not in the document]
Comment (fantes) [reply src/backends/torch/torchlib.cc:475]: remains of refactor , definitly erased

File: src/backends/torch/optim/ranger.h [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+  struct TORCH_API RangerParamState
+      : public torch::optim::OptimizerCloneableParamState<RangerParamState>
+  {
+    TORCH_ARG(int64_t, step) = 0;
+    TORCH_ARG(torch::Tensor, exp_avg);
+    TORCH_ARG(torch::Tensor, exp_avg_sq);
+    TORCH_ARG(torch::Tensor, max_exp_avg_sq) = {};
+    TORCH_ARG(torch::Tensor, slow_buffer);
+
+  public:
```
[the remark this answers is not in the document]
Comment (fantes) [reply src/backends/torch/optim/ranger.h:57]: ok

File: src/backends/torch/optim/ranger.h [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+    TORCH_ARG(double, weight_decay) = 0.0;
+    TORCH_ARG(bool, decoupled_wd) = false;
+    TORCH_ARG(bool, rectified) = true;
+    TORCH_ARG(bool, lookahead) = true;
+    TORCH_ARG(bool, adabelief) = false;
+    TORCH_ARG(bool, gradient_centralization) = false;
+    TORCH_ARG(int, lsteps) = 6;
+    TORCH_ARG(double, lalpha) = 0.5;
+
+  public:
```
[the remark this answers is not in the document]
Comment (fantes) [reply src/backends/torch/optim/ranger.h:40]: ok, removed

File: src/backends/torch/torchmodule.h (hunk 9 above)
Comment (fantes) [reply src/backends/torch/torchmodule.h:53]: removed this declaration from here :)

File: docs/api.md (hunk 3 above)
Comment (fantes) [reply docs/api.md:755]: ok done

File: src/backends/torch/optim/ranger.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (fantes) [reply src/backends/torch/optim/ranger.h:40]: indeed I put it back; stuff below is considered private by g++ (maybe due to the macro TORCH_API?)

File: src/backends/torch/optim/ranger.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (fantes) [reply src/backends/torch/optim/ranger.h:57]: ndeed I put it back; stuff below is considered private by g++ (maybe due to the macro TORCH_API?)

File: src/backends/torch/optim/ranger.h (hunk 6 above)
Comment (beniz) [reply src/backends/torch/optim/ranger.h:52]: OK, j etais sur que c etait active partout.. comme quoi...

File: src/backends/torch/optim/ranger.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (BynaryCobweb) [reply src/backends/torch/optim/ranger.h:78]: ok

**Review by beniz — APPROVED** [review]

**Comment by beniz** [comment]
image
